ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift

```diff
@@ -442,6 +442,10 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
         setupVerticalConstraints(textLabel)
         setupVerticalConstraints(learnMoreButton)
         
+        if let finalVerticalView = setupFinalVerticalViews() {
+            lastView = finalVerticalView
```

Comment:
For more context, this is my subclass in mPower app

open class SBATrackedMedicationNavigationHeaderView: RSDTableStepHeaderView {
    
    private var _underlinedButtonContraints: [NSLayoutConstraint] = []
    
    /// The label for displaying step detail text.
    @IBOutlet open var underlinedButton: RSDUnderlinedButton?
    
    public override init() {
        super.init(frame: CGRect.zero)
        commonInitNavigationHeaderView()
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        commonInitNavigationHeaderView()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInitNavigationHeaderView()
    }
    
    private func commonInitNavigationHeaderView() {
        addUnderlinedButtonIfNeeded()
    }
    
    override open func setupFinalVerticalViews() -> UIView? {
        // Remove existing constraints for the underlined button
        NSLayoutConstraint.deactivate(_underlinedButtonContraints)
        _underlinedButtonContraints.removeAll()
        
        // Add the underlined button under the title label
        if let underlinedButtonUnwrapped = self.underlinedButton,
            let titleLabelUnwrapped = super.titleLabel {
            _underlinedButtonContraints.append(contentsOf:
                underlinedButtonUnwrapped.rsd_alignBelow(view: titleLabelUnwrapped, padding: constants.verticalSpacing))
        }
        return underlinedButton
    }
    
    /// Convenience method for adding a label.
    open func addUnderlinedButton(font: UIFont, color: UIColor) -> RSDUnderlinedButton {
        let button = RSDUnderlinedButton(type: .system)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.titleLabel?.font = font
        button.setTitleColor(color, for: .normal)
        button.titleLabel?.textAlignment = .center
        button.titleLabel?.preferredMaxLayoutWidth = constants.labelMaxLayoutWidth
        self.addSubview(button)
        
        button.rsd_alignToSuperview([.leading, .trailing], padding: constants.sideMargin)
        button.rsd_makeHeight(.greaterThanOrEqual, 60.0)
        
        return button
    }
    
    func addUnderlinedButtonIfNeeded() {
        guard underlinedButton == nil else { return }
        underlinedButton = addUnderlinedButton(font: UIFont.rsd_headerTextLabel, color: UIColor.rsd_headerTextLabel)
    }
}

Reply:
Discussed in Slack:
themdp [5:31 PM]
[...]
the problem is, the header calculates it height through autolayout and it does that by setting the last view's bottom constraint to superview
so i needed a way to inject a view as the last view
in mpower, i did this...

themdp [5:33 PM]
added this Java snippet: Untitled 
open class SBATrackedMedicationNavigationHeaderView: RSDTableStepHeaderView {
  
  private var _underlinedButtonContraints: [NSLayoutConstraint] = []
  
  /// The label for displaying step detail text.
  @IBOutlet open var underlinedButton: RSDUnderlinedButton?
  
  public override init() {
    super.init(frame: CGRect.zero)
    commonInitNavigationHeaderView()
  }
  
  public override init(frame: CGRect) {
    super.init(frame: frame)
    commonInitNavigationHeaderView()
  }
  
  required public init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    commonInitNavigationHeaderView()
  }
  
  private func commonInitNavigationHeaderView() {
    addUnderlinedButtonIfNeeded()
  }
  
  override open func setupFinalVerticalViews() -> UIView? {
    // Remove existing constraints for the underlined button
    NSLayoutConstraint.deactivate(_underlinedButtonContraints)
    _underlinedButtonContraints.removeAll()
    
    // Add the underlined button under the title label
    if let underlinedButtonUnwrapped = self.underlinedButton,
      let titleLabelUnwrapped = super.titleLabel {
      _underlinedButtonContraints.append(contentsOf:
        underlinedButtonUnwrapped.rsd_alignBelow(view: titleLabelUnwrapped, padding: constants.verticalSpacing))
    }
    return underlinedButton
  }
  
  /// Convenience method for adding a label.
  open func addUnderlinedButton(font: UIFont, color: UIColor) -> RSDUnderlinedButton {
    let button = RSDUnderlinedButton(type: .system)
    button.translatesAutoresizingMaskIntoConstraints = false
    button.titleLabel?.font = font
    button.setTitleColor(color, for: .normal)
    button.titleLabel?.textAlignment = .center
    button.titleLabel?.preferredMaxLayoutWidth = constants.labelMaxLayoutWidth
    self.addSubview(button)
    
    button.rsd_alignToSuperview([.leading, .trailing], padding: constants.sideMargin)
    button.rsd_makeHeight(.greaterThanOrEqual, 60.0)
    
    return button
  }
  
  func addUnderlinedButtonIfNeeded() {
    guard underlinedButton == nil else { return }
    underlinedButton = addUnderlinedButton(font: UIFont.rsd_headerTextLabel, color: UIColor.rsd_headerTextLabel)
  }
}
Collapse 

themdp [5:34 PM]
i basically keep my own _interactiveConstraints as a var and keep track of those
sorry i should be posting this in the PR

erinmounts-sage [5:38 PM]
got it, that makes sense since it gets called whenever the updateInteractiveConstraints gets called

themdp [5:40 PM]
Based on the subclass context, is there still a more appropriate way I could go about this?

erinmounts-sage [5:41 PM]
I mean, if there’s a way to do it by overriding the public updateConstraints() func and calling super at the appropriate time, that would be ideal

themdp [5:43 PM]
That was my initial thought.  But, that would require me to somehow remove only the bottomConstraint of the lastView, which i would have no idea how to do.
does that make sense?

erinmounts-sage [5:47 PM]
I am trying to come up with a “cleaner” alternative and I think I am just recapitulating your thought process and coming to the same conclusions

themdp [5:47 PM]
lol
ive been there on a pr review
I think you were on to something though with your comment about setupVertical constraint
I think we could just put this in its own method
func setupVerticalConstraints(_ nextView: UIView?) {
           if let vw = nextView, shouldLayout(vw) {
               applyVerticalConstraint(to: vw, lastView: lastView)
               firstView = firstView == nil ? vw : firstView
               lastView = vw
           } else {
               nextView?.isHidden = true
           }
       }
       
       // image view
       setupVerticalConstraints(imageView)
       setupVerticalConstraints(titleLabel)
       setupVerticalConstraints(textLabel)
       setupVerticalConstraints(learnMoreButton)
and overriding that wouldnt accomplish my goal i think in a cleaner way
let me see if I can do it

erinmounts-sage [5:51 PM]
hm, yes that does feel a bit cleaner, and should be equivalent--please document that method as a suitable place to override to add more views to the vertical stack (seems like you should call super *after* injecting your views, but I’m not 100% positive about that)

themdp [5:52 PM]
The order of constraint additions doesnt matter, just their priority value
So I think you could do either way

erinmounts-sage [5:52 PM]
well since it depends on the value of lastView
and you’re changing lastView

themdp [5:53 PM]
Good thought, I'll look into that more